### PR TITLE
New version of ParallelFor to match RFC

### PR DESCRIPTION
**Pull request by jbruestle**

### CODE REVIEW

File: pmlc/dialect/pxa/ir/ops.td [hunk 1]
```diff
@@ -48,8 +48,17 @@ def ImplicitAffineTerminator
 
 def AffineParallelForOp : PXA_Op<"parallel_for", [ImplicitAffineTerminator]> {
   let summary = "multi-index parallel for operation";
-  let arguments = (ins I64ArrayAttr:$ranges, Variadic<Index>:$dynamic_ranges);
-  let regions = (region SizedRegion<1>:$inner);
+  let arguments = (ins AffineMapAttr:$ranges, AffineMapAttr:$transform, Variadic<Index>:$mapOperands);
+  let regions = (region SizedRegion<1>:$region);
+  let extraClassDeclaration = [{
+    ValueRange rangeMapOperands() { return llvm::make_range(operand_begin(), operand_begin() + ranges().getNumInputs()); }
+    ValueRange transformMapOperands() { return llvm::make_range(operand_begin() + ranges().getNumInputs(), operand_end()); }
```
Comment (flaub) [inline pmlc/dialect/pxa/ir/ops.td]: I think we should be consistent and use a `get` prefix for all of these.

File: pmlc/conversion/tile_to_pxa/tests/double_dot.mlir [hunk 2]
```diff
@@ -28,11 +28,9 @@ func @double_dot(
 // CHECK:   %2 = affine.load %arg1[%arg4, %arg6] : memref<20x30xf32>
 // CHECK:   %3 = mulf %1, %2 : f32
 // CHECK:   "pxa.reduce"(%3, %0, %arg4, %arg5, %arg6) {agg = 1 : i64, map = #map2} : (f32, memref<10x30xf32>, index, index, index) -> ()
-// CHECK: ranges = [20, 10, 30]
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tests/double_dot.mlir]: If we aren't checking for ranges here, perhaps we should check them elsewhere with a CHECK on the parallel for itself.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 3]
```diff
@@ -586,13 +580,13 @@ struct ContractionOpConversion : public OpConversionPattern<ContractionOp> {
     // TODO: addInitializer
 
     // Make the outer loops
-    auto dynamicRanges = ArrayRef<Value>();
-    auto forOp = rewriter.create<pxa::AffineParallelForOp>(loc, rewriter.getI64ArrayAttr(ranges), dynamicRanges);
-    auto body = rewriter.createBlock(&forOp.inner());
+    auto forOp = rewriter.create<pxa::AffineParallelForOp>(loc, ranges);
+    auto body = forOp.getBody();
+    rewriter.setInsertionPointToStart(body);
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: Can we use `rewriter.createBlock`? I think it internally sets the insertion point.

File: pmlc/dialect/pxa/ir/ops.cc [hunk 4]
```diff
@@ -1,10 +1,38 @@
 // Copyright 2019, Intel Corporation
 
+#include <iostream>
```
Comment (flaub) [inline pmlc/dialect/pxa/ir/ops.cc]: Remove this

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 5]
```diff
@@ -518,9 +515,6 @@ struct EltwiseOpConversion : public OpConversionPattern<FromOpType> {
     // Create the store
     rewriter.create<AffineStoreOp>(loc, result, resultMemRef, idxs);
 
-    // Terminate the inner body
-    rewriter.create<AffineTerminatorOp>(loc);
-
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: I think we might still want to keep this to make the dialect verify properly.

Comment (flaub) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: Oh, I see ensureTerminator in the builder, cool!

File: pmlc/conversion/tile_to_pxa/tests/double_dot.mlir (hunk 2 above)
Comment (jbruestle) [reply pmlc/conversion/tile_to_pxa/tests/double_dot.mlir]: The new results look like this:

#map0 = (d0, d1) -> (d1, d0)
#map1 = (d0, d1) -> (d0, d1)
#map2 = (d0, d1, d2) -> (d1, d2)
#map3 = () -> (20, 10, 30)
#map4 = (d0, d1, d2) -> (d0, d1, d2)
#map5 = () -> (30, 10, 40)


module {
  func @double_dot(%arg0: memref<10x20xf32>, %arg1: memref<20x30xf32>, %arg2: memref<30x40xf32>, %arg3: memref<10x40xf32>) {
    %0 = alloc() : memref<10x30xf32>
    "pxa.parallel_for"() ( {
    ^bb0(%arg4: index, %arg5: index, %arg6: index):	// no predecessors
      %1 = affine.load %arg0[%arg5, %arg4] : memref<10x20xf32>
      %2 = affine.load %arg1[%arg4, %arg6] : memref<20x30xf32>
      %3 = mulf %1, %2 : f32
      "pxa.reduce"(%3, %0, %arg4, %arg5, %arg6) {agg = 1 : i64, map = #map2} : (f32, memref<10x30xf32>, index, index, index) -> ()
      "affine.terminator"() : () -> ()
    }) {ranges = #map3, transform = #map4} : () -> ()
    "pxa.parallel_for"() ( {
    ^bb0(%arg4: index, %arg5: index, %arg6: index):	// no predecessors
      %1 = affine.load %0[%arg5, %arg4] : memref<10x30xf32>
      %2 = affine.load %arg2[%arg4, %arg6] : memref<30x40xf32>
      %3 = mulf %1, %2 : f32
      "pxa.reduce"(%3, %arg3, %arg4, %arg5, %arg6) {agg = 1 : i64, map = #map2} : (f32, memref<10x40xf32>, index, index, index) -> ()
      "affine.terminator"() : () -> ()
    }) {ranges = #map5, transform = #map4} : () -> ()
    return
  }
}

Not exactly sure how to capture the maps and them compare them.  Maybe CHECK-DAG of some sort?  Or maybe looks for the maps and get their names?

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 3 above)
Comment (jbruestle) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: The affine for already has a block, and createBlock was incorrectly adding another one.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 5 above)
Comment (jbruestle) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: The new constructor automatically does ensureTerminator, so adding this breaks things via two terminators.

File: pmlc/dialect/pxa/ir/ops.td (hunk 1 above)
Comment (jbruestle) [reply pmlc/dialect/pxa/ir/ops.td]: Done.

File: pmlc/conversion/tile_to_pxa/tests/double_dot.mlir (hunk 2 above)
Comment (flaub) [reply pmlc/conversion/tile_to_pxa/tests/double_dot.mlir]: I think check for maps then check that their name is used (using regex capture probably)

**Pull request merged**